train.py

```diff
@@ -0,0 +1,183 @@
+
+
+def intersection_over_union(input: torch.Tensor, target: torch.Tensor, threshold: float = 0.5):
+    # input and target are probabilities of belonging to certain class
+    pred = input > threshold
+    mask = target > threshold
+
+    iou = (pred & mask).sum() / (pred | mask).sum()
+
+    return iou
+
+
+class UnetTrainingWrapper(pl.LightningModule):
+    def __init__(self, model: Unet, lr: float, weight_decay: float):
+        super().__init__()
+
+        self.model = model
+
+        self.save_hyperparameters(ignore=[model])
+
+        # metrics
+        self.loss_fn = nn.BCEWithLogitsLoss()
+        self.accuracy = torchmetrics.Accuracy(task="binary")
+        self.f1_score = torchmetrics.F1Score("binary")
+
+    def forward(self, x: torch.Tensor):
+        return self.model(x)
+
+    def configure_optimizers(self) -> optim.Optimizer:
+        optimizer = optim.AdamW(self.model.parameters(), lr=self.hparams["lr"], weight_decay=self.hparams["weight_decay"])
+
+        return optimizer
+
+    def _step(self, batch: tuple[torch.Tensor, torch.Tensor], batch_idx: int):
+        # extract signal and mask from batch
+        signal, mask = batch
+
+        mask_logits = self.model(signal)
+
+        # calculate loss
+        loss = self.loss_fn(mask_logits, mask)
+
+        # mask probabilities, used for calculating accuracy and f1 score
+        mask_pred = torch.sigmoid(mask_logits)
+        # calculate other metrics
+        acc = self.accuracy(mask_pred, mask)
+        f1 = self.f1_score(mask_pred, mask)
+        iou = intersection_over_union(mask_pred, mask)
+
+        return loss, acc, f1, iou
+
+    def training_step(self, batch: tuple[torch.Tensor, torch.Tensor], batch_idx: int):
+        loss, acc, f1, iou = self._step(batch, batch_idx)
+
+        self.log_dict({"train/loss": loss, "train/accuracy": acc, "train/f1-score": f1, "train/iou": iou}, on_epoch=True)
+
+        return loss
+
+    def validation_step(self, batch: tuple[torch.Tensor, torch.Tensor], batch_idx: int):
+        loss, acc, f1, iou = self._step(batch, batch_idx)
+
+        self.log_dict({"val/loss": loss, "val/accuracy": acc, "val/f1-score": f1, "train/iou": iou}, on_epoch=True)
+
+    def test_step(self, batch: tuple[torch.Tensor, torch.Tensor], batch_idx: int):
+        loss, acc, f1, iou = self._step(batch, batch_idx)
+
+        self.log_dict({"test/loss": loss, "test/accuracy": acc, "test/f1-score": f1, "train/iou": iou}, on_epoch=True)
+
+
+def makedir_if_not_exists(dir: str):
+    if not os.path.exists(dir):
+        os.makedirs(dir)
+
+
+def preprocess_dataset(dataset_name: str, batch_size: int, num_workers: int, seed: int = 0):
+    dataset = ECGDataset(dataset_name)
+
+    # generator used to determine which records belong to which split
+    generator = torch.Generator().manual_seed(seed)
+
+    # spliting data based on seed
+    train_ds, val_ds, test_ds = random_split(dataset, [0.8, 0.1, 0.1], generator=generator)
```

Comment:
I think this will mix data from all the records together - my idea was to have the split defined by `record_id`, so something like this:

```python
ltafdb_records= ["100", ...]
train_records, test_records = random_split(ltafdb_records)
train_ds = Dataset(train_records)
test_ds = Dataset(test_records)
```

Let me know if this makes sense. Splitting the existing HF dataset by record id may be tricky, so maybe it would be better to create splits before pushing to hub

Reply:
Yeah, probably it was better to do them before pushing to hub